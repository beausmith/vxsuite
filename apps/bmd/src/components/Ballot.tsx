import React, { useContext } from 'react'
import { BrowserRouter, Redirect, Route, Switch } from 'react-router-dom'

import BallotContext from '../contexts/ballotContext'

import Screen from '../components/Screen'
import ContestPage from '../pages/ContestPage'
import StartPage from '../pages/StartPage'
import SummaryPage from '../pages/SummaryPage'

const Ballot = () => {
  const { contests } = useContext(BallotContext)
  return (
    <BrowserRouter>
      <Screen>
        <Switch>
          <Route path="/" exact component={StartPage} />
          <Redirect exact from="/contests" to={`/contests/${contests[0].id}`} />
          <Route path="/contests/:id" component={ContestPage} />
          <Route path="/summary" component={SummaryPage} />
        </Switch>
      </Screen>
    </BrowserRouter>
  )
}

export default Ballot
