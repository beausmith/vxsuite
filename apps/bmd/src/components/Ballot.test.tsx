// Accessibility is not tested because ballot alone does not render html.
// Each route component is tested elsewhere.

import lodashMerge from 'lodash.merge'
import React from 'react'
import { fireEvent } from 'react-testing-library'

import { render } from '../../test/testUtils'

import electionSample from '../data/electionSample.json'

import Ballot from './Ballot'

it(`can navigate all ballot pages`, () => {
  const { container, getByText, getByTestId } = render(<Ballot />)
  expect(container.firstChild).toMatchSnapshot()

  fireEvent.change(getByTestId('activation-code'), {
    target: { value: 'MyVoiceIsMyPassword' },
  })
  // TODO: replace next line with "Enter" keyDown on activation code input
  fireEvent.click(getByText('Submit'))
  fireEvent.click(getByText('Get Started'))
  fireEvent.click(getByText('Next'))
  fireEvent.click(getByText('Help'))
  getByText('Help content will be available here.')
  fireEvent.click(getByText('Back'))
  fireEvent.click(getByText('Settings'))
  getByText('Settings will be available here.')
  fireEvent.click(getByText('Back'))
  fireEvent.click(getByText('Review'))
  fireEvent.click(getByText('Back'))
})

it(`redirects contests index to first contest`, () => {
  const { getByText } = render(<Ballot />, {
    route: '/contests',
  })
  getByText(electionSample.contests[0].title)
})

it(`skips activation page if disabled`, () => {
  const { getByText } = render(<Ballot />, {
    election: lodashMerge(electionSample, {
      bmdConfig: {
        requireActivation: false,
      },
    }),
    route: '/',
  })
  getByText('Get Started')
})
