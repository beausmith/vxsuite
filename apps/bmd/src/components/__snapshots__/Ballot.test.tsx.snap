// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can navigate all ballot pages 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
  padding: 1rem 0.5rem 2rem;
}

.c1 {
  max-width: 35rem;
  margin: auto auto;
}

.c3 {
  line-height: 1.2;
  max-width: 66ch;
  text-align: center;
  margin: auto;
}

.c3 h1 {
  font-size: 1.5rem;
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c3 h2 {
  font-size: 1.25rem;
  margin: 1.5rem 0 0.75rem;
}

.c3 h3,
.c3 p {
  font-size: 1rem;
  margin: 1rem 0;
}

.c3 h1 + h2 {
  margin-top: -0.75rem;
}

.c3 h1 + p,
.c3 h2 + p {
  margin-top: -0.75rem;
}

.c3 h3 + p {
  margin-top: -1rem;
}

.c3 :first-child {
  margin-top: 0;
}

.c3 :last-child {
  margin-bottom: 0;
}

.c5 > svg {
  display: block;
  height: auto;
  width: 100%;
}

.c2 {
  margin: auto;
}

.c4 {
  border: 1rem solid white;
  max-width: 320px;
  margin: auto;
}

@media (min-width:480px) {
  .c0 {
    padding: 2rem 1rem;
  }
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media print {
  .c1 {
    max-width: 100%;
    margin: 0;
  }
}

@media (min-width:480px) {
  .c3 {
    line-height: 1.3;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <form>
      <fieldset>
        <legend
          class="c2"
        >
          <label
            for="BallotCode"
          >
            <div
              class="c3"
            >
              <h1>
                Scan Your Activation Code
              </h1>
              <p>
                Your ballot will be displayed after scan is complete.
              </p>
              <div
                class="c4"
              >
                <div
                  class="c5"
                >
                  <svg
                    height="128"
                    shape-rendering="crispEdges"
                    viewBox="0 0 25 25"
                    width="128"
                  >
                    <path
                      d="M0,0 h25v25H0z"
                      fill="#FFFFFF"
                    />
                    <path
                      d="M0 0h7v1H0zM8 0h3v1H8zM12 0h2v1H12zM16 0h1v1H16zM18,0 h7v1H18zM0 1h1v1H0zM6 1h1v1H6zM10 1h3v1H10zM16 1h1v1H16zM18 1h1v1H18zM24,1 h1v1H24zM0 2h1v1H0zM2 2h3v1H2zM6 2h1v1H6zM9 2h1v1H9zM13 2h1v1H13zM15 2h2v1H15zM18 2h1v1H18zM20 2h3v1H20zM24,2 h1v1H24zM0 3h1v1H0zM2 3h3v1H2zM6 3h1v1H6zM8 3h2v1H8zM11 3h1v1H11zM13 3h1v1H13zM18 3h1v1H18zM20 3h3v1H20zM24,3 h1v1H24zM0 4h1v1H0zM2 4h3v1H2zM6 4h1v1H6zM9 4h2v1H9zM12 4h1v1H12zM18 4h1v1H18zM20 4h3v1H20zM24,4 h1v1H24zM0 5h1v1H0zM6 5h1v1H6zM9 5h2v1H9zM13 5h1v1H13zM16 5h1v1H16zM18 5h1v1H18zM24,5 h1v1H24zM0 6h7v1H0zM8 6h1v1H8zM10 6h1v1H10zM12 6h1v1H12zM14 6h1v1H14zM16 6h1v1H16zM18,6 h7v1H18zM10 7h3v1H10zM15 7h1v1H15zM2 8h1v1H2zM4 8h3v1H4zM8 8h3v1H8zM12 8h2v1H12zM15 8h1v1H15zM17 8h1v1H17zM21 8h1v1H21zM24,8 h1v1H24zM0 9h1v1H0zM2 9h3v1H2zM10 9h1v1H10zM12 9h2v1H12zM15 9h3v1H15zM20 9h1v1H20zM22,9 h3v1H22zM0 10h1v1H0zM4 10h1v1H4zM6 10h1v1H6zM9 10h1v1H9zM11 10h1v1H11zM18 10h1v1H18zM24,10 h1v1H24zM0 11h2v1H0zM5 11h1v1H5zM9 11h4v1H9zM15 11h2v1H15zM20 11h2v1H20zM23 11h1v1H23zM2 12h1v1H2zM5 12h2v1H5zM10 12h1v1H10zM12 12h2v1H12zM18 12h1v1H18zM20 12h2v1H20zM2 13h2v1H2zM7 13h3v1H7zM13 13h1v1H13zM15 13h1v1H15zM19 13h1v1H19zM21 13h1v1H21zM0 14h1v1H0zM4 14h3v1H4zM8 14h2v1H8zM13 14h2v1H13zM17 14h1v1H17zM23,14 h2v1H23zM1 15h4v1H1zM7 15h3v1H7zM12 15h4v1H12zM17 15h3v1H17zM21 15h1v1H21zM23,15 h2v1H23zM0 16h1v1H0zM3 16h2v1H3zM6 16h2v1H6zM10 16h1v1H10zM12 16h2v1H12zM16 16h6v1H16zM23,16 h2v1H23zM8 17h2v1H8zM11 17h1v1H11zM14 17h3v1H14zM20 17h2v1H20zM0 18h7v1H0zM11 18h1v1H11zM14 18h1v1H14zM16 18h1v1H16zM18 18h1v1H18zM20,18 h5v1H20zM0 19h1v1H0zM6 19h1v1H6zM8 19h1v1H8zM11 19h4v1H11zM16 19h1v1H16zM20 19h2v1H20zM23 19h1v1H23zM0 20h1v1H0zM2 20h3v1H2zM6 20h1v1H6zM8 20h3v1H8zM12 20h9v1H12zM0 21h1v1H0zM2 21h3v1H2zM6 21h1v1H6zM9 21h2v1H9zM15 21h2v1H15zM18 21h1v1H18zM22 21h2v1H22zM0 22h1v1H0zM2 22h3v1H2zM6 22h1v1H6zM8 22h4v1H8zM13 22h3v1H13zM18 22h3v1H18zM24,22 h1v1H24zM0 23h1v1H0zM6 23h1v1H6zM12 23h3v1H12zM17 23h1v1H17zM21 23h1v1H21zM23 23h1v1H23zM0 24h7v1H0zM9 24h4v1H9zM14 24h1v1H14zM16 24h6v1H16zM23,24 h2v1H23z"
                      fill="#000000"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </label>
        </legend>
        <input
          aria-hidden="true"
          autocapitalize="off"
          autocomplete="off"
          autocorrect="off"
          class="visually-hidden"
          data-testid="activation-code"
          id="BallotCode"
          spellcheck="false"
          type="text"
          value=""
        />
        <button
          aria-hidden="true"
          class="visually-hidden"
          type="submit"
        >
          Submit
        </button>
      </fieldset>
    </form>
  </div>
</main>
`;
